cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,699 @@
+                                         int64_t>(builder, batch, offset, length);
+      case liborc::BYTE:
+        return append_numeric_batch_cast<Int8Builder, int8_t, liborc::LongVectorBatch,
+                                         int64_t>(builder, batch, offset, length);
+      case liborc::DOUBLE:
+        return append_numeric_batch<DoubleBuilder, liborc::DoubleVectorBatch, double>(
+            builder, batch, offset, length);
+      case liborc::FLOAT:
+        return append_numeric_batch_cast<FloatBuilder, float, liborc::DoubleVectorBatch,
+                                         double>(builder, batch, offset, length);
+      case liborc::BOOLEAN:
+        return append_bool_batch(builder, batch, offset, length);
+      case liborc::VARCHAR:
+      case liborc::STRING:
+        return append_binary_batch<StringBuilder>(builder, batch, offset, length);
+      case liborc::BINARY:
+        return append_binary_batch<BinaryBuilder>(builder, batch, offset, length);
+      case liborc::CHAR:
+        return append_fixed_binary_batch(builder, batch, offset, length);
+      case liborc::DATE:
+        return append_date_batch(builder, batch, offset, length);
+      case liborc::TIMESTAMP:
+        return append_timestamp_batch(builder, batch, offset, length);
+      case liborc::DECIMAL:
+        return append_decimal_batch(type, builder, batch, offset, length);
+      default:
+        std::stringstream ss;
+        ss << "Not implemented type kind: " << kind;
+        return Status::NotImplemented(ss.str());
+    }
+  }
+
+  Status append_struct_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                             liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                             int64_t length) {
+    auto builder = static_cast<StructBuilder*>(abuilder);
+    auto batch = static_cast<liborc::StructVectorBatch*>(cbatch);
+
+    const uint8_t* valid_bytes = nullptr;
+    if (batch->hasNulls) {
+      valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+    }
+    RETURN_NOT_OK(builder->Append(length, valid_bytes));
+
+    int n_fields = builder->num_fields();
+    for (int i = 0; i < n_fields; i++) {
+      RETURN_NOT_OK(append_batch(type->getSubtype(i), builder->field_builder(i),
+                                 batch->fields[i], offset, length));
+    }
+    return Status::OK();
+  }
+
+  Status append_list_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                           liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                           int64_t length) {
+    auto builder = static_cast<ListBuilder*>(abuilder);
+    auto batch = static_cast<liborc::ListVectorBatch*>(cbatch);
+    liborc::ColumnVectorBatch* elements = batch->elements.get();
+    const liborc::Type* elemtype = type->getSubtype(0);
+
+    for (int i = offset; i < length + offset; i++) {
+      if (!batch->hasNulls || batch->notNull[i]) {
```

Comment:
If you set `const bool nullable = batch->hasNulls` then some compilers may be able to unswitch this loop. Not sure it will matter much performancewise though